### PR TITLE
Add Rush repository configuration and use it in "rush change" to find…

### PR DESCRIPTION
**Pull request by qz2017**
… the source branch to compare with.

### CODE REVIEW

File: apps/rush-lib/src/rush.schema.json [hunk 1]
```diff
@@ -70,6 +70,17 @@
       },
       "additionalProperties": false
     },
+    "repository": {
+      "description": "The repository location",
+      "type": "object",
+      "properties": {
+        "url": {
+          "description": "The url of the repository",
```
Comment (pgonzal) [inline apps/rush-lib/src/rush.schema.json]: >"The url of the repository" [](start = 25, length = 27)

This documentation is not very helpful -- obviously the "repoistory.url" stores the URL of the repository.  :-)

Maybe you could explain:
- Why do we need this URL?
- When is it NOT needed?
- Is there any special format required?
- What happens if it is not provided?

File: apps/rush-lib/src/utilities/VersionControl.ts [hunk 2]
```diff
@@ -41,6 +41,41 @@ export default class VersionControl {
     });
   }
 
+  public static getRemoteMasterBranch(repositoryUrl?: string): string | undefined {
```
Comment (pgonzal) [inline apps/rush-lib/src/utilities/VersionControl.ts]: >getRemoteMasterBranch [](start = 16, length = 21)

Please add some documentation explaining what this function is doing...  it is not obvious

File: apps/rush-lib/src/utilities/VersionControl.ts [hunk 3]
```diff
@@ -41,6 +41,41 @@ export default class VersionControl {
...
+          return remoteUrl === repositoryUrl;
+        }
+        return false;
+      });
+    }
+
+    if (useDefault) {
+      return defaultMaster;
+    } else if (matchingRemotes.length > 0) {
+      return `${matchingRemotes[0]}/master`;
```
Comment (pgonzal) [inline apps/rush-lib/src/utilities/VersionControl.ts]: >return `${matchingRemotes[0]}/master`; [](start = 6, length = 38)

Is this common?  If not, should we issue a warning?

File: apps/rush-lib/src/utilities/VersionControl.ts [hunk 4]
```diff
@@ -41,6 +41,41 @@ export default class VersionControl {
...
+          }
+          return remoteUrl === repositoryUrl;
+        }
+        return false;
+      });
+    }
+
+    if (useDefault) {
+      return defaultMaster;
+    } else if (matchingRemotes.length > 0) {
```
Comment (pgonzal) [inline apps/rush-lib/src/utilities/VersionControl.ts]: >matchingRemotes.length > 0 [](start = 15, length = 26)

Is it common to have more than one matching remote?  If not, should we issue a warning?

File: apps/rush-lib/src/data/RushConfiguration.ts [hunk 5]
```diff
@@ -58,13 +58,24 @@ export interface IEventHooksJson {
 }
 
 /**
+ * Part of IRushConfigurationJson.
+ */
+export interface IRushRepositoryJson {
+  /**
+   * Url of the repository
```
Comment (pgonzal) [inline apps/rush-lib/src/data/RushConfiguration.ts]: >Url of the repository [](start = 5, length = 21)

(See my other note about documenting this)

File: apps/rush-lib/src/utilities/VersionControl.ts [hunk 6]
```diff
@@ -41,6 +41,41 @@ export default class VersionControl {
...
+    } else {
+      const output: string = child_process
+      .execSync(`git remote`)
+      .toString();
+      matchingRemotes = output.split('\n').filter(remoteName => {
+        if (remoteName) {
+          const remoteUrl: string = child_process.execSync(`git remote get-url ${remoteName}`)
+            .toString()
+            .trim();
+          if (remoteName === defaultRemote && remoteUrl === repositoryUrl) {
```
Comment (pgonzal) [inline apps/rush-lib/src/utilities/VersionControl.ts]: >remoteUrl === repositoryUrl [](start = 46, length = 27)

should this be a case-insensitive comparison?

Comment (qz2017) [inline apps/rush-lib/src/utilities/VersionControl.ts]: It is safer to compare url in a case-sensitive way.

---
In reply to: [134301332](https://github.com/Microsoft/web-build-tools/pull/328#discussion_r134301332) [](ancestors = 134301332)

**Pull request merged**